### PR TITLE
Table Groups

### PR DESCRIPTION
**Pull request by IsaacCloos**
#44 

| Checklist  | status |
| --- | :---: |
| Code  | ✅  |
| Documentation  | ✅  |
| Example File | ✅ |
| Tests | ✅ |
| Readme | ✅ |

# Criticism Welcome

### CODE REVIEW

File: src/display/macros.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+/// |                                     | ├─────────┼──────────┤ |
+/// |                                     | │ Manjaro │ Arch     │ |
+/// |                                     | └─────────┴──────────┘ |
+/// +-------------------------------------+------------------------+
+/// */
+/// ```
+///
+/// ### Clean Display
+///
+/// The `@` identifier is recognized by all patterns to produce
```
Comment (zhiburt) [inline src/display/macros.rs:82]: Personally if I'd use the library I'd take `@` to be something a bit strange :)

I think it's not nessary?
I mean if someone would like to make it `clean`  he/she could just use `with.(Style::empty())`.
Am I wrong here?

Comment (IsaacCloos) [reply src/display/macros.rs:82]: I completely agree.

I will *clean* this up 👍🏻

**Comment by zhiburt** [comment]
It looks legit.

I am just thinking for a few hours whether it's that valuable.
I mean someone could just do what macro does `[table1.to_string(), table2.to_string()]` (the same for chunch and cloned version).
Maybe I am missing something what do you think?

But I am thinking maybe what be better is if the macro could do column build too.
I am not sure how though.

But if we could build table like this.

```rust
// build table as a row
table![t1, t2]

// build table as a column
table![
   // build table as a row
   table![t1, t2],
   
   // build table as a row
   table![t3, t4],
]

// thinking....
// build table as a column like it also be great......
table![t1, t2]
```

We could introduce 2 macros 1 for rows 1 for columns which would work.

```rust
// build table as a row
row![t1, t2]

// build table as a column
column![
   // some table
   t1,

   // build table as a row
   row![t1, t2],
   
   // build table as a row
   row![t3, t4],
]

column![t1, t2]
```

Just thinking..................

ref: #44

File: src/display/macros.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+/// |                                     |                        |
+/// |                                     |                        |
+/// |                                     |                        |
+/// +-------------------------------------+------------------------+
+/// */
+/// ```
+#[macro_export]
+macro_rules! group {
+    // Parallel Display
+    ( $($table:expr), * ) => {{
```
Comment (zhiburt) [inline src/display/macros.rs:88]: You only need to add a support for trailing comma.
See

```rust
let table = group![
   Table::new(&data),
   Table::new(&data),
];
```

**Review by zhiburt — COMMENTED** [review]
It looks legit.

I just can't stop thinking whether it's valuable.
I mean someone could just do what macro does `[table1.to_string(), table2.to_string()]` (the same for chunch and cloned version).
Maybe I am missing something what do you think?

But I am thinking maybe what be better is if the macro could do column build too.
I am not sure how though.

But if we could build table like this.

```rust
// build table as a row
table![t1, t2]

// build table as a column
table![
   // build table as a row
   table![t1, t2],
   
   // build table as a row
   table![t3, t4],
]

// thinking....
// build table as a column like it also be great......
table![t1, t2]
```

__________________________

We can introduce 2 macros 1 for rows 1 for columns which works.

```rust
row![t1, t2]

column![
   t1,
   row![t1, t2],
   row![t3, t4],
]

column![t1, t2]
```

(`row!` is literally what you've implemented already)

Just thinking..................

ref: #44

**Comment by zhiburt** [comment]
@IsaacCloos so how do you feel about my comment?

If you don't have time we can merge it and iterate on it.

**Comment by IsaacCloos** [comment]
Many interesting thoughts here. I will try to address them all.

### Equivalence to native functions

> I mean someone could just do what macro does [table1.to_string(), table2.to_string()] (the same for chunk and cloned version).

Here are some examples of what you describe that I could think of:


```rust
Table::new([table_a.to_string(), table_b.to_string()])
```

```text
+---------------------------------------------+
| String                                      |
+---------------------------------------------+
| +-------+-----+--------------+              |
| | name  | age | is_validated |              |
| +-------+-----+--------------+              |
| | Sam   | 31  | true         |              |
| +-------+-----+--------------+              |
| | Sarah | 26  | true         |              |
| +-------+-----+--------------+              |
+---------------------------------------------+
| ┌────────────────────┬─────┬──────────────┐ |
| │ name               │ age │ is_validated │ |
| ├────────────────────┼─────┼──────────────┤ |
| │ Jack Black         │ 51  │ false        │ |
| ├────────────────────┼─────┼──────────────┤ |
| │ Michelle Goldstein │ 44  │ true         │ |
| └────────────────────┴─────┴──────────────┘ |
+---------------------------------------------+
```

```rust
Table::new([
    [table_a.to_string(), table_b.to_string()],
    [table_b.to_string(), table_c.to_string()],
])
```

```text
+---------------------------------------------+---------------------------------------------+
| 0                                           | 1                                           |
+---------------------------------------------+---------------------------------------------+
| +-------+-----+--------------+              | ┌────────────────────┬─────┬──────────────┐ |
| | name  | age | is_validated |              | │ name               │ age │ is_validated │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sam   | 31  | true         |              | │ Jack Black         │ 51  │ false        │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sarah | 26  | true         |              | │ Michelle Goldstein │ 44  │ true         │ |
| +-------+-----+--------------+              | └────────────────────┴─────┴──────────────┘ |
+---------------------------------------------+---------------------------------------------+
| ┌────────────────────┬─────┬──────────────┐ | .---------------------------------------.   |
| │ name               │ age │ is_validated │ | | name             | age | is_validated |   |
| ├────────────────────┼─────┼──────────────┤ | | Jon Doe          | 255 | false        |   |
| │ Jack Black         │ 51  │ false        │ | | Mark Nelson      | 13  | true         |   |
| ├────────────────────┼─────┼──────────────┤ | | Terminal Monitor | 0   | false        |   |
| │ Michelle Goldstein │ 44  │ true         │ | | Adam Blend       | 17  | true         |   |
| └────────────────────┴─────┴──────────────┘ | '---------------------------------------'   |
+---------------------------------------------+---------------------------------------------+
```

```rust
Table::new([
    [table_a.to_string(), table_b.to_string()],
    [table_b.to_string(), Default::default()],
])
```

```text
+---------------------------------------------+---------------------------------------------+
| 0                                           | 1                                           |
+---------------------------------------------+---------------------------------------------+
| +-------+-----+--------------+              | ┌────────────────────┬─────┬──────────────┐ |
| | name  | age | is_validated |              | │ name               │ age │ is_validated │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sam   | 31  | true         |              | │ Jack Black         │ 51  │ false        │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sarah | 26  | true         |              | │ Michelle Goldstein │ 44  │ true         │ |
| +-------+-----+--------------+              | └────────────────────┴─────┴──────────────┘ |
+---------------------------------------------+---------------------------------------------+
| ┌────────────────────┬─────┬──────────────┐ |                                             |
| │ name               │ age │ is_validated │ |                                             |
| ├────────────────────┼─────┼──────────────┤ |                                             |
| │ Jack Black         │ 51  │ false        │ |                                             |
| ├────────────────────┼─────┼──────────────┤ |                                             |
| │ Michelle Goldstein │ 44  │ true         │ |                                             |
| └────────────────────┴─────┴──────────────┘ |                                             |
+---------------------------------------------+---------------------------------------------+
```

I believe a macro would improve the user experience of writing these examples.

### table! Concepts

Since `group!` produces a table, and seems to be a convenient way to build tables. I can see the value of renaming it `table!`.

Brainstorm:

```rust
table![
    table![t1, t2],
    table![t3, t4],
]
```

It is my opinion that this feels a bit confusing compared to something like this:

```rust
table![
    [t1, t2],
    [t3, t4]
]
```

### row! and col!

```rust
column![
   t1,
   row![t1, t2],
   row![t3, t4],
]
```

are you thinking this would produce:

```text
+---------------------------------------------+---------------------------------------------+
| 0                                                                                         | <-- I'm not sure yet what would be done about the columns in this example
+---------------------------------------------+---------------------------------------------+
|                               +-------+-----+--------------+                              |
|                               | name  | age | is_validated |                              |
|                               +-------+-----+--------------+                              |
|                               | Sam   | 31  | true         |                              |
|                               +-------+-----+--------------+                              |
|                               | Sarah | 26  | true         |                              |
|                               +-------+-----+--------------+                              |
+---------------------------------------------+---------------------------------------------+
| +-------+-----+--------------+              | ┌────────────────────┬─────┬──────────────┐ |
| | name  | age | is_validated |              | │ name               │ age │ is_validated │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sam   | 31  | true         |              | │ Jack Black         │ 51  │ false        │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sarah | 26  | true         |              | │ Michelle Goldstein │ 44  │ true         │ |
| +-------+-----+--------------+              | └────────────────────┴─────┴──────────────┘ |
+---------------------------------------------+---------------------------------------------+
| ┌────────────────────┬─────┬──────────────┐ | .---------------------------------------.   |
| │ name               │ age │ is_validated │ | | name             | age | is_validated |   |
| ├────────────────────┼─────┼──────────────┤ | | Jon Doe          | 255 | false        |   |
| │ Jack Black         │ 51  │ false        │ | | Mark Nelson      | 13  | true         |   |
| ├────────────────────┼─────┼──────────────┤ | | Terminal Monitor | 0   | false        |   |
| │ Michelle Goldstein │ 44  │ true         │ | | Adam Blend       | 17  | true         |   |
| └────────────────────┴─────┴──────────────┘ | '---------------------------------------'   |
+---------------------------------------------+---------------------------------------------+
```

or this:

```text
+---------------------------------------------+---------------------------------------------+
| 0                                           | 1                                           |
+---------------------------------------------+---------------------------------------------+
| +-------+-----+--------------+              |                                             |
| | name  | age | is_validated |              |                                             |
| +-------+-----+--------------+              |                                             |
| | Sam   | 31  | true         |              |                                             |
| +-------+-----+--------------+              |                                             |
| | Sarah | 26  | true         |              |                                             |
| +-------+-----+--------------+              |                                             |
+---------------------------------------------+---------------------------------------------+
| +-------+-----+--------------+              | ┌────────────────────┬─────┬──────────────┐ |
| | name  | age | is_validated |              | │ name               │ age │ is_validated │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sam   | 31  | true         |              | │ Jack Black         │ 51  │ false        │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sarah | 26  | true         |              | │ Michelle Goldstein │ 44  │ true         │ |
| +-------+-----+--------------+              | └────────────────────┴─────┴──────────────┘ |
+---------------------------------------------+---------------------------------------------+
| ┌────────────────────┬─────┬──────────────┐ | .---------------------------------------.   |
| │ name               │ age │ is_validated │ | | name             | age | is_validated |   |
| ├────────────────────┼─────┼──────────────┤ | | Jon Doe          | 255 | false        |   |
| │ Jack Black         │ 51  │ false        │ | | Mark Nelson      | 13  | true         |   |
| ├────────────────────┼─────┼──────────────┤ | | Terminal Monitor | 0   | false        |   |
| │ Michelle Goldstein │ 44  │ true         │ | | Adam Blend       | 17  | true         |   |
| └────────────────────┴─────┴──────────────┘ | '---------------------------------------'   |
+---------------------------------------------+---------------------------------------------+
```

Is the idea that this would produce 1 column not 2?

### Sorry for huge post 😅 

I hope that addressed all of your thoughts. If you have specific requests for `group!` I'd be happy to do what I can 👍🏻

**Comment by zhiburt** [comment]
> I believe `group!` improves the user experience of writing these examples.

Correct.
Except that you can't do that using only `Table::new` nor only `group`.
(I think you can't)

> Since group! produces a table, and seems to be a convenient way to build tables. I can see the value of renaming it table!.

The same from me.
I just don't think we can cover creation of column/row in 1 macro.
So I think to rename it to `row!` and a similar macro for `column!` need to be created.


> Is the idea that this would produce 1 column not 2?

I was thinking differently.
See this examples.

```rust
row![table1, table2]
```

```
+---------------------------------------------+---------------------------------------------+
| +-------+-----+--------------+              | ┌────────────────────┬─────┬──────────────┐ |
| | name  | age | is_validated |              | │ name               │ age │ is_validated │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sam   | 31  | true         |              | │ Jack Black         │ 51  │ false        │ |
| +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ |
| | Sarah | 26  | true         |              | │ Michelle Goldstein │ 44  │ true         │ |
| +-------+-----+--------------+              | └────────────────────┴─────┴──────────────┘ |
+---------------------------------------------+---------------------------------------------+
```

```rust
column![table1, table2]
```

```
+---------------------------------------------+
| +-------+-----+--------------+              |
| | name  | age | is_validated |              |
| +-------+-----+--------------+              |
| | Sam   | 31  | true         |              |
| +-------+-----+--------------+              |
| | Sarah | 26  | true         |              |
| +-------+-----+--------------+              |
+---------------------------------------------+
| ┌────────────────────┬─────┬──────────────┐ |
| │ name               │ age │ is_validated │ |
| ├────────────────────┼─────┼──────────────┤ |
| │ Jack Black         │ 51  │ false        │ |
| ├────────────────────┼─────┼──────────────┤ |
| │ Michelle Goldstein │ 44  │ true         │ |
| └────────────────────┴─────┴──────────────┘ |
+---------------------------------------------+
```

```rust
column![
   t1,
   row![t1, t2],
]
```

```
+-----------------------------------------------------------------------------------------------+
| +-------+-----+--------------+                                                                |
| | name  | age | is_validated |                                                                |
| +-------+-----+--------------+                                                                |
| | Sam   | 31  | true         |                                                                |
| +-------+-----+--------------+                                                                |
| | Sarah | 26  | true         |                                                                |
| +-------+-----+--------------+                                                                |
+-----------------------------------------------------------------------------------------------+
| +---------------------------------------------+---------------------------------------------+ |
| | +-------+-----+--------------+              | ┌────────────────────┬─────┬──────────────┐ | |
| | | name  | age | is_validated |              | │ name               │ age │ is_validated │ | |
| | +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ | |
| | | Sam   | 31  | true         |              | │ Jack Black         │ 51  │ false        │ | |
| | +-------+-----+--------------+              | ├────────────────────┼─────┼──────────────┤ | |
| | | Sarah | 26  | true         |              | │ Michelle Goldstein │ 44  │ true         │ | |
| | +-------+-----+--------------+              | └────────────────────┴─────┴──────────────┘ | | 
| +---------------------------------------------+---------------------------------------------+ |
+-----------------------------------------------------------------------------------------------+
```

_______________________

So everything is the same as you did with `group!`.

`row!` returns `Table` and takes a list of `ToString`s.
`table!` returns `Table` and takes a list of `ToString`s.

How do you feel about it?

PS: Yes in case of inner tables there often `with(Style::empty)` be called but I still think sticking to `Style::ascii` is better. But I might be wrong.

**Comment by IsaacCloos** [comment]
Let me know what you think 😄

File: src/display/macros.rs [hunk 3]
```diff
@@ -101,25 +36,41 @@ macro_rules! group {
...
+///     table1,
+///     row![table2, table3]
+/// ];
+/// ```
+#[macro_export]
+macro_rules! col {
+    // Vertical
+    ( $($table:expr), * $(,)? ) => {{
+        let mut builder = Table::builder([ $($table.to_string(),)* ]);
+        builder.remove_columns();
```
Comment (zhiburt) [inline src/display/macros.rs:63]: (can be ignored)

I think creating `Builder::default` and then calling `add_record` be more efficient.
But that's OK.

**Review by zhiburt — COMMENTED** [review]
Look great :100: 

A few notes:

1. Split `row` and `col` macros by files?
2. Create a feature `macros` to hide the macros (not everyone will use them).
     Like here https://github.com/zhiburt/tabled/blob/fc14dc1cd857780f26e1b9e0b479057504f7e418/Cargo.toml#L31
https://github.com/zhiburt/tabled/blob/fc14dc1cd857780f26e1b9e0b479057504f7e418/src/lib.rs#L244-L245

File: examples/README.md [hunk 4]
```diff
@@ -318,7 +318,7 @@ vvvvvvvvvvvvvvvvvv
 +-----------------------------------------------+
 ```
 
-## group
+## col_row
```
[the remark this answers is not in the document]
Comment (IsaacCloos) [reply examples/README.md:321]: I thought the examples/README file was only the outputs? You can see the code examples in the examples/col_row.rs

Is that what you mean?

[the remark this answers is not in the document]
Comment (zhiburt) [reply examples/README.md]: Yes but short example how you obtain it be cool to have.
Just as you did.

**Review by zhiburt — APPROVED** [review]
LGTM

Thanks for your contribution.
I'll merge it today.

close #44

**Pull request merged**